### PR TITLE
GSK-844 Push feature

### PR DESCRIPTION
**Pull request by Googleton**
## Description

Implementation of the Push feature

### CODE REVIEW

File: backend/src/main/java/ai/giskard/service/InspectionService.java [hunk 1]
```diff
@@ -235,4 +238,16 @@ public Inspection updateInspectionName(long inspectionId, String name) {
...
+//            ExplainRequest request = ExplainRequest.newBuilder()
+//                .setModel(grpcMapper.createRef(model))
+//                .setDataset(grpcMapper.createRef(dataset))
+//                .putAllColumns(Maps.filterValues(features, Objects::nonNull))
+//                .build();
+//
+//            return client.getBlockingStub().explain(request);
+//        }
+//    }
+
```
Comment (Googleton) [inline backend/src/main/java/ai/giskard/service/InspectionService.java:252]:
```suggestion
```

**Comment by andreybavt** [comment]
I'm getting an error when no internal worker is connected. @Googleton I don't understand why we're always using an internal worker for pushes generation, it should be project dependant, otherwise you won't be able to execute a model requiring custom python dependencies
Screenshot 2023-05-22 at 12 21 57

**Comment by andreybavt** [comment]
z-index issue

image

File: python-client/giskard/push/transformation.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        return text.upper()
+
+    def lower_case(self, text):
+        return text.lower()
+
+    def title_case(self, text):
+        return text.title()
+
+
+text = "She her the brown fox jumps over the lazy dog in Sydney, was christian"
```
Comment (andreybavt) [inline python-client/giskard/push/transformation.py:117]: this should be a unit test

File: python-client/tests/test_push.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+def test_contribution(german_credit_model, german_credit_data):
+    testl = []
+    for i in range(100):
+        res = contribution(german_credit_model, german_credit_data, i)
+        if res is not None:
+            testl.append(res)
+    assert len(testl) > 0
+
+
+# def test_perturbation_text(enron_model, enron_data):
```
Comment (andreybavt) [inline python-client/tests/test_push.py:23]: cleanup here and below

**Review by andreybavt — CHANGES_REQUESTED** [review]

File: python-client/giskard/push/transformation.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
+import random
+
+import spacy
+
+# Load the English language model
+nlp = spacy.load('en_core_web_sm')
```
Comment (mattbit) [inline python-client/giskard/push/transformation.py:6]: We don’t want to do this here!!!

File: python-client/giskard/push/transformation.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+
+class TextTransformer:
+    def __init__(self):
+        self.nlp = spacy.load("en_core_web_sm")
```
Comment (mattbit) [inline python-client/giskard/push/transformation.py:11]: What if this fails? This requires the user to have installed the extra spacy module `en_core_web_sm`.

File: python-client/giskard/push/perturbation.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        probabilities = model.model.predict(row_perturbed)  # .reshape(1, -1)
+        return Perturbation(passed=ref_prob[0] != probabilities[0], perturbation_value=perturbation_val)
+    elif model.meta.model_type == SupportedModelTypes.REGRESSION:
+        ref_val = model.model.predict(ref_row.drop(columns=[ds.target]))
+        new_val = model.model.predict(row_perturbed.drop(columns=[ds.target]))  # .reshape(1, -1)
+        return Perturbation(passed=(new_val - ref_val) / ref_val >= 0.2, perturbation_value=perturbation_val)
+
+
+def _num_perturb(val, mad):
+    return val + 3 * mad  # 1.2  # 20% perturbation
```
Comment (mattbit) [inline python-client/giskard/push/perturbation.py:87]: This is roughly 2 σ for a normal distribution. To me it looks like a big perturbation! Are we sure we want this? E.g. if value is close to the mean, we would be pushing it to the 95% percentile.

Comment (mattbit) [reply python-client/giskard/push/perturbation.py:87]: Also: we only increase the value! Perturbation could be in the negative direction. (Maybe I am missing something?)

File: python-client/giskard/push/transformation.py
Comment (mattbit) [inline python-client/giskard/push/transformation.py:1]: Quite an overlap with scan perturbation, can we merge the common parts?

**Review by mattbit — CHANGES_REQUESTED** [review]
A few things to check on the Python part:
- `Push` class might need to be refactored in a cleaner way
- Dependency with `spacy`: do we need this? do we commit to adding this to the dependencies?
- Perturbation-based push has a significant overlap with the scan feature: can we isolate the transformations and reuse them in both push and scan?

File: python-client/giskard/ml_worker/server/ml_worker_service.py [hunk 7]
```diff
@@ -477,6 +477,52 @@ def filter_wrapper(row):
...
+        except ModuleNotFoundError as e:
+            raise GiskardException(
+                f"Failed to import '{e.name}'. "
+                f"Make sure it's installed in the ML Worker environment."
+                "To have more information on ML Worker, please see: https://docs.giskard.ai/start/guides/installation/ml-worker"
+            ) from e
+        from giskard.push.contribution import contribution
+        from giskard.push.perturbation import perturbation
+
+        contribs = contribution(model, dataset, request.rowidx)
```
[the remark this answers is not in the document]
Comment (mathieuroques) [reply python-client/giskard/ml_worker/server/ml_worker_service.py:503]: Done

File: python-client/giskard/push/contribution.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                    res = Push(push_type="contribution_only", feature=el,
+                               value=values[el],
+                               bounds=bounds
+                               )
+                    yield res
+
+    if model.meta.model_type == SupportedModelTypes.REGRESSION:
+        shap_res = _contribution_push(model, ds, idrow)
+        values = ds.df.iloc[idrow]
+        # re = ds.__dict__
```
[the remark this answers is not in the document]
Comment (mathieuroques) [reply python-client/giskard/push/contribution.py:45]: Done

File: python-client/giskard/push/contribution.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+from ..push import Push
+from .utils import slice_bounds
+
+
+def contribution(model, ds, idrow):  # data_aug_dict
+    if model.meta.model_type == SupportedModelTypes.CLASSIFICATION:
+        shap_res = _contribution_push(model, ds, idrow)
+        values = ds.df.iloc[idrow]
+        training_label = values[ds.target]
+        prediction = model.model.predict(ds.df.iloc[[idrow]])
```
[the remark this answers is not in the document]
Comment (mathieuroques) [reply python-client/giskard/push/contribution.py]: Resolved

File: python-client/giskard/push/transformation.py (hunk 2 above)
Comment (mathieuroques) [reply python-client/giskard/push/transformation.py:117]: I removed this unused file

File: python-client/tests/test_push.py (hunk 3 above)
Comment (mathieuroques) [reply python-client/tests/test_push.py]: Done

File: python-client/giskard/push/transformation.py (hunk 4 above)
Comment (mathieuroques) [reply python-client/giskard/push/transformation.py:6]: File deleted

File: python-client/giskard/push/transformation.py
Comment (mathieuroques) [reply python-client/giskard/push/transformation.py:1]: File deleted

**Comment by andreybavt** [comment]
Closing this PR as @Googleton will open a new one for the branch where the conflicts are merged. We're using this branch as a reference for conflict resolution

**Pull request closed without merging**